---
title: "Differential Expression Analysis: Comparisons between KO versus WT in timepoints 0, 6, 12 and 24 hours"
author:
- name: Computational Immunogenomics Core (CIG Core)
  affiliation: Center for Systems Immunology, University of Pittsburgh
  email: obe3@pitt.edu
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true # table of content true
    toc_depth: 3  # upto three depths of headings (specified by #, ## and ###)
    toc_float:
      collapsed: true
      smooth_scroll: true
    number_sections: true  ## if you want number sections at each table header
    theme: "default" # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(biomaRt)
library(sleuth)
library(openxlsx)
library(ggplot2)
library(ggfortify)
library(dplyr)
library(matrixStats)
library(pheatmap)
library(VennDiagram)
library(stringr)
library(ReactomePA)
library(clusterProfiler)
library(WebGestaltR)
library(org.Mm.eg.db)
library(AnnotationDbi)
library(enrichplot)
library(tidyverse)

source("/ix/cigcore/utils/code/sleuth.func.R")

# Set paths
proj_path <- "/ix/cigcore/proj/sarkar"
data_path <- file.path(proj_path, "data")
result_path <- file.path(proj_path, "result/01.RawData")
experiment <- "09-01-2023-DE_multivariateAnalysis"
experiment_path <- file.path(result_path, experiment)
dir.create(experiment_path, recursive=TRUE)

```


```{r sleuth-setup, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
# Inputs for sleuth.func
dataDir <- file.path(result_path, "mapped_kallisto")
metadataPath <- file.path(data_path, "IRF1-KO_CD8_Tcells_Metadata.csv")

skipLines <- 0
sampleCol <- "ID"

factorCol <- c("Cells", "Treatment_hr")
factorOfInterest <- "Cells"

refLevel <- c("WT", 0)
outPath <- experiment_path

mart <- useMart("ensembl", dataset="mmusculus_gene_ensembl") 
ttg <- getBM(attributes=c("ensembl_gene_id", "ensembl_transcript_id", "external_gene_name"), mart = mart) 
colnames(ttg) <- c("ensembl_gene", "target_id", "gene_name")
#write.csv(ttg, file.path(experiment_path, "ttg.csv"), row.names = FALSE)
aggCol <- "ensembl_gene"

pvalueCutoff <- 0.05 # used when performing pathway analysis 
qval <- 0.05 # used for differential expression analysis
bval <- 0
```

```{r run_sleuth.func, echo=FALSE, warning=FALSE, message=FALSE, include=FALSE}
sleuth_results <- sleuth.func(metaData = metadataPath, skipLines = skipLines, sampleCol = sampleCol, factors = factorCol, 
                                factorOfInt = factorOfInterest, refLevel = refLevel, dataDir = dataDir, mappingDF = ttg, 
                                aggCol = aggCol, pvalCutoff = qval, outPath = outPath, drop_dupGenes = FALSE)
save(sleuth_results, file = file.path(outPath, "sleuth_results.RData")) # saves R object list
```

# Sample Metadata

```{r metadata, echo=FALSE}
sampleInfo <- rbind(sleuth_results$sampleDF)
numSamples <- dim(sampleInfo)[1]
knitr::kable(sampleInfo)
```

## Summary on Tools used for this Analysis 

**Transcript-level quantification with Kallisito**

[Kallisto](https://www.nature.com/articles/nbt.3519) is an alignment free tool that quantifies RNA-seq data on the transcript (isoform) level. It utilizes a pseudo alignment approach which rapidly determines the compatibility of reads with targets (transcript sequences). 

**Differential expression with Sleuth**

[Sleuth](https://www.nature.com/articles/nmeth.4324) is a tool used for differential analysis of gene expression for which transcript abundances have been quantified with kallisto.

Test used for differential expression analysis is **Wald Test (WT)** is used to assess the significance of the difference in gene expression levels between two or more experimental conditions. The test is applied to each gene/transcript to determine whether the difference in expression levels is statistically significant or not.

## Comparisons of Interest

In order to determine which pairs are significantly different from each other, we carried out multiple comparisons. This comparisons are written as *variable group vs. reference group*. 

* **KO vs. WT (for all timepoints)**

## Sample Quality Check: Principal Component Analysis (PCA)

Sample clustering is checked with principal component analysis.


### PCA Plot for Comparisons of Interest

**KO vs. WT**

```{r pca_1, echo=FALSE}
sleuth_results$KO_vs_WT_pca_Trans
```

### Differential Expression Results

Number of samples: `r numSamples`

*To identify significant differential expressed transcripts, we used this cutoff below:*

q-value (Adjusted p-value): `r qval`

b-value (log2 fold-change): `r bval`

**Summary Table** 

This summary table is showing the number of significant differential expressed transcripts for each comparisons of interest. 

```{r all_summary, echo=FALSE}
summaryTable <- sleuth_results$summaryDF
summaryTable <- dplyr::select(summaryTable, Comparisons, DE_Genes, DE_Transcripts, Up_Reg_Transcripts, Down_Reg_Transcripts) # for sigDET
#summaryTable <- dplyr::select(summaryTable, Comparisons, DE_Genes, DE_Transcripts, Up_Reg_Genes, Down_Reg_Genes) # for sigDEG
knitr::kable(summaryTable)
```

#### Volcano Plot

Note: **Reference level used for all comparisons is `r refLevel`**

**KO vs. WT**

```{r vPlot_1, echo=FALSE, warning=FALSE, message=FALSE}
sleuth_results$KO_vs_WT_vPlot_Trans
```

## Pathway Analysis

The false discovery rate (fdr) method used is **Benjamini-Hochberg (BH) adjustment**.

The adjusted pvalue cutoff used for these pathway analysis `r pvalueCutoff`. 

### Over-representation analysis (ORA)

**Over-Representation Analysis (ORA)** is a method for identifying biological pathways, gene sets, or other functional categories that are over represented in a list of differentially expressed genes, compared to what would be expected by chance. ORA is a widely used method in bioinformatics and functional genomics, and can help to identify biological pathways and processes that are dysregulated in a disease or condition of interest.

```{r run_ORA, include=FALSE, echo=FALSE, warning=FALSE, message=FALSE}
# Pathway analysis using clusterprofiler 
cp_path <- file.path(experiment_path, "pathway_analysis_ORA")
dir.create(cp_path, recursive=TRUE)

ORA <- list()
plots_list <- list()


sigDE_UpOrDn <- grep("Up|Dn", names(sleuth_results), value=TRUE)
sigDE_UpOrDn <- grep("^sigDET",sigDE_UpOrDn, value=TRUE) #change to sigDET or sigDEG if using transcript or genes to run ORA
for(g1 in sigDE_UpOrDn){
  # Get gene names and convert to entrez id
  genes <- sleuth_results[[g1]]$gene_name
  entrez_genes <- bitr(genes, fromType = "SYMBOL", toType = "ENTREZID", OrgDb = "org.Mm.eg.db")

  # Reactome pathway over-representation analysis
  reactome <- enrichPathway(gene = entrez_genes$ENTREZID, organism = "mouse", pvalueCutoff=pvalueCutoff, readable=TRUE)
  ORA[[paste0(g1, "_Reactome")]] <- reactome
  
  # KEGG pathway over-representation analysis
  kegg <- enrichKEGG(gene = entrez_genes$ENTREZID, organism = "mmu",  pvalueCutoff=pvalueCutoff, use_internal_data = FALSE)
  kegg2 <- setReadable(kegg, OrgDb = "org.Mm.eg.db", keyType="ENTREZID")
  kegg2@result$Description <- gsub(pattern = " - Mus musculus (house mouse)", replacement = "", kegg2@result$Description, fixed = T)    
  ORA[[paste0(g1, "_KEGG")]] <- kegg2
}


for(a1 in names(ORA)){
  reacORkegg <- ORA[[a1]]
  if(nrow(reacORkegg) !=0){
    # Create directories for each result
    ORA_path <- file.path(cp_path, a1)
    dir.create(ORA_path, recursive=TRUE)
    
    ORA_df <- as.data.frame(reacORkegg)
    write.csv(ORA_df, file.path(ORA_path, paste0(a1, ".csv")))
    
    # Plot enrichment result - dotplot
    fit <- plot(dotplot(reacORkegg, showCategory = 40))
    png(filename = file.path(ORA_path, paste0(a1, "_dotplot.png")), res = 300, width = 17, height = 10, units = "in")
    print(fit)
    dev.off()
    plots_list[[a1]] <- fit
  }
  
  if (nrow(reacORkegg) > 40){
    # Weighted Set Cover of geneSets
    weightedPath <- file.path(cp_path, "weightedSetCover", a1)
    dir.create(weightedPath, recursive=TRUE)
    
    setCoverNum = abs(0.40*(nrow(ORA_df)))
    nThreads = 4
    idsInSet <- sapply(ORA_df$geneID, strsplit, split="/")
    names(idsInSet) <- ORA_df$ID    
    minusLogP <- -log(ORA_df$pvalue)
    minusLogP[minusLogP == Inf] <- -log(.Machine$double.eps)
    wscRes <- weightedSetCover(idsInSet=idsInSet, costs=(1 / minusLogP), topN=setCoverNum, nThreads=nThreads)

    wscRes_full <- ORA_df[c(match(wscRes$topSets, ORA_df$ID)),]
    wscRes_full <- wscRes_full[order(wscRes_full$p.adjust), ]
    write.csv(wscRes_full, file.path(weightedPath, paste0(a1, ".csv")))
    
    # Plot enrichment result of weighted set cover - dotplot
    reacORkegg2 <- reacORkegg
    reacORkegg2@result <- wscRes_full
      
    fit <- plot(dotplot(reacORkegg2, showCategory = 40))
    png(filename = file.path(weightedPath, paste0(a1, "_dotplot.png")), res = 300, width = 17, height = 10, units = "in")
    print(fit)
    dev.off()
    plots_list[[paste0(a1, "weightSet")]] <- fit
  }
}
```

## Reference Level: WT

**KO vs. WT - Up-regulated (KEGG)**

```{r enrichment_plot_1-KEGG-Up, echo=FALSE, warning=FALSE, message=FALSE, fig.width=15, fig.height=12}
plots_list$sigDET_KO_vs_WT_Up_KEGGweightSet
```

**KO vs. WT - Down-regulated (KEGG)**

```{r enrichment_plot_2-KEGG-Dn, echo=FALSE, warning=FALSE, message=FALSE, fig.width=15, fig.height=10}
plots_list$sigDET_KO_vs_WT_Dn_KEGGweightSet
```

**KO vs. WT - Up-regulated (Reactome)**

```{r enrichment_plot_3-Reactome-Up, echo=FALSE, message=FALSE, warning=FALSE, fig.width=15, fig.height=12}
plots_list$sigDET_KO_vs_WT_Up_Reactome
```

**KO vs. WT - Down-regulated (Reactome)**

```{r enrichment_plot_4-Reactome-Dn, echo=FALSE, warning=FALSE, message=FALSE, fig.width=15, fig.height=13}
plots_list$sigDET_KO_vs_WT_Dn_ReactomeweightSet
```

## Gene Set Enrichment Analysis(GSEA)

**Gene Set Enrichment Analysis (GSEA)** is a computational method for analyzing gene expression data that seeks to identify statistically significant differences in the expression of predefined sets of genes, or gene sets, between two or more biological conditions. GSEA is widely used in pathway analysis to identify biological pathways or gene sets that are differential regulated between groups of samples.

```{r run_gsea, include=FALSE, echo=FALSE, warning=FALSE, message=FALSE}
cp_path <- file.path(experiment_path, "pathway_analysis_GSEA")
dir.create(cp_path, recursive=TRUE)

GSEA <- list()
plots_list <- list()

allDE_list <- grep("allDET", names(sleuth_results), value=TRUE) #change to allDET or allDEG_bvalue if using transcript or genes to run GSEA
for(g1 in  allDE_list){ 
  # Remove duplicated gene names 
  allDET <- sleuth_results[[g1]]
  allDET_sorted <- allDET[order(allDET$ensembl_gene, -allDET$b),]
  allDET_sorted <- allDET_sorted[!duplicated(allDET_sorted$ensembl_gene),]
 
  # Get gene names and convert to entrez id
  genes <- allDET_sorted$gene_name
  entrez_genes <- bitr(genes, fromType = "SYMBOL", toType = "ENTREZID", OrgDb = "org.Mm.eg.db")
  genes_df <- merge(allDET_sorted, entrez_genes, by.x = "gene_name", by.y ="SYMBOL")
  
  # Get gene lists
  gene_list <- genes_df$b
  names(gene_list) <- genes_df$ENTREZID
  gene_list = sort(gene_list, decreasing = TRUE)
  
  # Reactome pathway over-representation analysis
  reactome <- gsePathway(gene = gene_list, organism = "mouse", pvalueCutoff=pvalueCutoff)
  reactome2 <- setReadable(reactome, OrgDb = "org.Mm.eg.db", keyType="ENTREZID")
  GSEA[[paste0(g1, "_Reactome")]] <- reactome2
  
  # KEGG pathway over-representation analysis
  kegg <- gseKEGG(gene = gene_list, organism = "mmu", use_internal_data = FALSE, pvalueCutoff=pvalueCutoff)
  kegg2 <- setReadable(kegg, OrgDb = "org.Mm.eg.db", keyType="ENTREZID")
  kegg2@result$Description <- gsub(pattern = " - Mus musculus (house mouse)", replacement = "", 
                                  kegg2@result$Description, fixed = T)    
  GSEA[[paste0(g1, "_KEGG")]] <- kegg2
}


for(a1 in names(GSEA)){
  reacORkegg <- GSEA[[a1]]
  if(nrow(reacORkegg) !=0){
    # Create directories for each result
    GSEA_path <- file.path(cp_path, a1)
    dir.create(GSEA_path, recursive=TRUE)
    
    GSEA_df <- as.data.frame(reacORkegg)
    write.csv(GSEA_df, file.path(GSEA_path, paste0(a1, ".csv")))
    
    # Plot enrichment result - dotplot
    fit <- plot(dotplot(reacORkegg, x= "NES", showCategory = 40))
    png(filename = file.path(GSEA_path, paste0(a1, "_dotplot.png")), res = 300, width = 17, height = 10, units = "in")
    print(fit)
    dev.off()
    plots_list[[a1]] <- fit
  }
  
  if(nrow(reacORkegg) > 40){
    # Weighted Set Cover of geneSets
    weightedPath <- file.path(cp_path, "weightedSetCover", a1)
    dir.create(weightedPath, recursive=TRUE)
    
    setCoverNum = abs(0.40*(nrow(GSEA_df)))
    nThreads = 4
    idsInSet <- sapply(GSEA_df$core_enrichment, strsplit, split="/")
    names(idsInSet) <- GSEA_df$ID    
    minusLogP <- -log(GSEA_df$pvalue)
    minusLogP[minusLogP == Inf] <- -log(.Machine$double.eps)
    wscRes <- weightedSetCover(idsInSet=idsInSet, costs=(1 / minusLogP), topN=setCoverNum, nThreads=nThreads)
    
    wscRes_full <- GSEA_df[c(match(wscRes$topSets, GSEA_df$ID)),]
    wscRes_full <- wscRes_full[order(wscRes_full$p.adjust), ]
    write.csv(wscRes_full, file.path(weightedPath, paste0(a1, ".csv")))
    
    # Plot enrichment result of weighted set cover - dotplot
    reacORkegg2 <- reacORkegg
    reacORkegg2@result <- wscRes_full
      
    fit <- plot(dotplot(reacORkegg2, x= "NES", showCategory = 40))
    png(filename = file.path(weightedPath, paste0(a1, "_dotplot.png")), res = 300, width = 17, height = 10, units = "in")
    print(fit)
    dev.off()
    plots_list[[paste0(a1, "weightSet")]] <- fit
  }
}
```

## Reference Level: WT

**KO vs. WT - Reactome**

```{r enrichment_plot_1-Reactome, echo=FALSE, message=FALSE, warning=FALSE, fig.width=12, fig.height=7}
plots_list$KO_vs_WT_allDET_Reactome
```

**KO vs. WT - KEGG**

```{r enrichment_plot_2-KEGG, echo=FALSE, message=FALSE, warning=FALSE}
plots_list$KO_vs_WT_allDET_KEGG
```

### Enrichment Plot for selected pathways from GSEA

```{r enrichmentplot, include=FALSE, echo=FALSE, warning=FALSE, message=FALSE}
enrichmentPath <- file.path(experiment_path,  "forPublication/pathway_analysis_GSEA/enrichnmentPlot")
dir.create(enrichmentPath, recursive=TRUE)

interestedPathways <- c("Cytokine Signaling in Immune system", "Cytokine-cytokine receptor interaction", "TCR signaling", 
                        "Downstream TCR signaling", "T cell receptor signaling pathway")
for(g1 in names(GSEA)){
  gseaPathways <- GSEA[[g1]]
  gseaPath <- file.path(enrichmentPath, g1)
  dir.create(gseaPath, recursive=TRUE)
  
  for(i1 in interestedPathways){
    geneSetID <- which(gseaPathways$Description == i1)
    if(length(geneSetID) > 0){
      title <- gseaPathways$Description[geneSetID]
      enrichmentPlot <- gseaplot2(gseaPathways, geneSetID =  geneSetID, title = title)
      png(filename = file.path(gseaPath, paste0(i1, ".png")))
      print(enrichmentPlot)
      dev.off()
    }
  }
  if(length(list.files(gseaPath)) == 0) {
  unlink(gseaPath, recursive = TRUE)
  } 
}
```

## Description of Files and its Contents

These files are found in the shared One drive folder. The contents of each files are described below: 

* **sig_DE.xlsx:**These excel files are named as *"sig_DE_variable_vs_reference.xlsx"*.
    + **sigDET:** 
        - *target_id:* Ensembl transcript identifier 
        - *pval:* p-value for each transcripts in the analysis 
        - *qval:* p-value adjusted for multiple testing by means of false discovery rate (FDR), using Benjamini-Hochberg
        - *b:* 'beta' value (effect size). Technically a biased estimator of the fold change. It is the log2 fold changes between conditions b/c we specified log2 transformation. 
        - *se_b:* standard error of the beta value
        - *mean_obs:* the mean expression (log2) of the transcript across all samples in the analysis
        - *var_obs:* the observed variance (biological variance) of the expression across all samples in the analysis
        - *tech_var:* the technical variance of expression (derived from the bootstraps)
        - *sigma_sq:* raw estimator of the variance once the technical variance has been removed
        - *smooth_sigma_sq:* the smooth regression fit for the shrinkage estimation
        - *final_sigma_sq:* max(sigma_sq, smooth_sigma_sq). used for covariance estimation of beta (in addition to tech_var)
    + **sigDEG:**
        - *num_aggregated_transcripts:* the number of transcripts that were aggregated to generate the expression estimate for each gene.The expression estimates for multiple transcripts corresponding to the same gene are combined to generate a single estimate for the gene. 
        - *pval and qval:* as explained above
        - *sum_mean_obs_counts:* the mean observed counts for all transcripts (across all samples in the analysis) that were aggregated to generate each gene-level expression estimate
    + **sigDET_Up/sigDEG_Up:** the positive values refers to *up regulated in the variable group* or *down regulated in reference group*
    + **sigDET_Dn/sigDEG_Dn:** the negative values refers to *down regulated in the variable group* or *up regulated in reference group*
    
* **TPM.csv:** contains the TPM (Transcript per million) values. It accounts for sequencing depth and gene length. 
* **pathway_analysis_GSEA & pathway_analysis_ORA:** each folder contains pathway analyses results of comparisons of interest using Gene Set Enrichment Analysis (GSEA) and Over-Representation Analysis (ORA) respectively

